### ISSUE TITLE
Support event batching in `demo_logs` source

### ISSUE THREAD

**Issue by JeanMertz** [vectordotdev/vector#12504]
### A note for the community

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Use Cases

The `demo_logs` source, [as the name implies](https://github.com/vectordotdev/vector/pull/9979), is intended for demo'ing Vector using random logs generated by Vector itself. It is **not** intended for benchmarking.

Yet, try as we might, people (including ourselves) are still using it for quick benchmarking. The most recent example is [here](https://github.com/vectordotdev/vector/issues/12481).

One of the problems seen in that discussion is that the demo logs source does not support batching, it sends one event per "batch".

We're entering a gray area here on what we should and should not support for the demo logs source, but I would propose that showing _somewhat_ realistic performance numbers is also important in a demo environment, especially ones we don't control or see, and I assume that even if we don't use it in our soak tests, many users/companies use it as a quick way to get some raw performance numbers, as they are evaluating the use of Vector.

As noted in #9979, making this source a "full benchmarking-capable source" requires significant effort, but I would propose that adding a feature to this source to allow it to batch events, would make it at least somewhat more realistic as a quick benchmarking tool, the way people are already using it.

**Comment by hhromic** [comment]
> The `demo_logs` source, [as the name implies](https://github.com/vectordotdev/vector/pull/9979), is intended for demo'ing Vector using random logs generated by Vector itself. It is **not** intended for benchmarking.

That's a very interesting discussion that clarifies a lot of wrong assumptions I personally had with the `demo_logs` source.
We have been using `demo_logs` for benchmarking and microbenchmarking and never realised it was not really suitable.

Same for the `blackhole` sink, we often used it as a terminal "no-op" for benchmarking other parts of the topology. After some discussion/enlightenment on Discord, we now started using the `http` sink with a dummy HTTP server instead.

We also use the `stdin` and `exec` sources for more realistic benchmarks because we use them in real-world topologies. After this recent discussion here and on Discord, I'm wondering if these sources are already as efficient as they could be. I'm not an expert in Vector's codebase but looks like `exec` is using event batching indeed but not sure about `stdin`.

The lesson learned for us is (for now) to stay away from `demo_logs` and `blackhole` for proper benchmarking. I can provide more details about how/why we do benchmarks with Vector if you are interested. 👍

**Comment by JeanMertz** [comment]
> One of the biggest reasons why we used the `demo_logs` source and `blackhole` sink is because they are very convenient and quick to setup for microbenchmarking of particular transforms or even particular functions in VRL. There is no need to setup dummy HTTP servers nor to setup an exec command (or something else) to obtain input data.

Agreed. This is also why we did the rename, to give more clarity to the purpose of the source. Still, for every person that takes part in our GitHub/Discord discussions, and understands how to do proper benchmarking, there are dozens of people using Vector who aren't aware of those nuances, and I therefore think it makes sense to make their experience as representative as possible, without any massive investment into a proper benchmarking tool built into Vector.

**Comment by hhromic** [comment]
> and I therefore think it makes sense to make their experience as representative as possible, without any massive investment into a proper benchmarking tool built into Vector.

Agreed. If the cost is not that big, definitively having "toy" sources that behave as close as possible to more "serious" sources, then I also think is a nice effort worth pursuing.
